File: graphs/findIfPathExistsInGraphDFS.cpp
class Solution {
public:
    bool dfs(unordered_map<int, vector<int>>& graph, unordered_map<int, bool>& visited, int current, int destination){
        if(current == destination){
            return true;
        }
        visited[current] = true;
        for(int neighbor: graph[current]){
            if(!visited[neighbor] && dfs(graph, visited, neighbor, destination)){
                return true;
            }
        }
        return false;
    }
    bool validPath(int n, vector<vector<int>>& edges, int source, int destination) {
        unordered_map<int, vector<int>> graph;
        unordered_map<int, bool> visited(n);
        for(const auto& edge: edges){
            int u = edge[0];
            int v = edge[1];
            graph[u].push_back(v);
            graph[v].push_back(u);
        } 
        return dfs(graph, visited, source, destination); 
    }
};

File: graphs/findIfPathExistsInGraphDFS.py
class Solution:
    def validPath(self, n: int, edges: List[List[int]], source: int, destination: int) -> bool:
        # create an adjacency list
        # starting at src, do a dfs 
        graph = defaultdict(list)
        for edge in edges:
            u, v = edge[0], edge[1]
            graph[u].append(v)
            graph[v].append(u)
        
        visit = [False for _ in range(n)]
        
        def dfs(cur):
            # base cases
            if cur == destination:
                return True
            if visit[cur]:
                return False
            
            # recursive cases
            visit[cur] = True
            for nei in graph[cur]:
                if dfs(nei):
                    return True
            return False
        return dfs(source)

        